extensions/amp-lightbox-viewer/0.1/amp-lightbox-viewer.js

```diff
@@ -0,0 +1,344 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {CSS} from '../../../build/amp-lightbox-viewer-0.1.css';
+import {isExperimentOn} from '../../../src/experiments';
+import {Layout} from '../../../src/layout';
+import {lightboxManagerForDoc} from '../../../src/lightbox-manager';
+import {user, dev} from '../../../src/log';
+import {ancestorElements} from '../../../src/dom';
+
+/** @const */
+const EXPERIMENT = 'amp-lightbox-viewer';
+
+class AmpLightboxViewer extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.NODISPLAY;
+  }
+
+  /** @override */
+  renderOutsideViewport() {
+    return true;
+  }
+
+  /** @override */
+  buildCallback() {
+    user().assert(isExperimentOn(this.win, EXPERIMENT),
+        `Experiment ${EXPERIMENT} disabled`);
+
+    /** @private {boolean} */
+    this.active_ = false;
+
+    /** @private {Element} */
+    this.activeElem_ = null;
+
+    /** @private {function(!Event)} */
+    this.boundHandleKeyboardEvents_ = this.handleKeyboardEvents_.bind(this);
+
+    /**
+     * @const
+     * @private {!../../../src/service/lightbox-manager-impl.LightboxManager}
+     */
+    this.manager_ = lightboxManagerForDoc(this.win.document.documentElement);
+
+    /** @const @private {Element} */
+    this.container_ = this.win.document.createElement('div');
+    this.container_.classList.add('-amp-lightbox-viewer');
+    this.buildMask_();
+    this.buildControls_();
+    this.element.appendChild(this.container_);
+
+    this.registerAction('open', this.activate.bind(this));
+  }
+
+  /** @override */
+  layoutCallback() {
+    // DO NOT ADD CODE HERE
+    // layoutCallback for lightbox-viewer is meaningless, lightbox-viewer
+    // doesn't have children, it just manages elements elsewhere in the page in
+    // `open_` `close_` and `updateViewer_` methods.
+  }
+
+  /**
+   * Builds the page mask and appends it to the container.
+   * @private
+   */
+  buildMask_() {
+    dev().assert(this.container_);
```

Comment:
I have added tests which should cover this too but I like the assert though, without it it would fail at `this.container_.appendChild(mask);` anyway but with assert at least it fails earlier.